<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced Options • QCpG</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced Options">
<meta property="og:description" content="A description including examples of the available advanced options in the QC app.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QCpG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AdvancedOptions.html">Advanced Options</a>
    </li>
    <li>
      <a href="../articles/AvailableMetrics.html">Available metrics</a>
    </li>
    <li>
      <a href="../articles/EWASworkflow.html">Example EWAS workflow</a>
    </li>
    <li>
      <a href="../articles/GetExampleData.html">Get example data</a>
    </li>
    <li>
      <a href="../articles/GettingStarted.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/UsingTheApp.html">Using the QC app</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="AdvancedOptions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Advanced Options</h1>
            
      

      <div class="hidden name"><code>AdvancedOptions.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="genotype-data-relatedness">Genotype data &amp; relatedness<a class="anchor" aria-label="anchor" href="#genotype-data-relatedness"></a>
</h2>
<div class="section level3">
<h3 id="include-genotype-data-to-detect-sample-swapscontamination">Include genotype data to detect sample swaps/contamination<a class="anchor" aria-label="anchor" href="#include-genotype-data-to-detect-sample-swapscontamination"></a>
</h3>
<p>The 450k/EPIC arrays contain a set (±60) of probes designed to measure common genetic variants By default, within the <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function relatedness info (identity by state, IBS) is inferred from the DNA methylation using these SNPs. Users can also provide genotype data to check if the samples match the respective genetic data, which is useful for detecting sample swaps/contamination.</p>
<p>The <a href="https://bioconductor.org/packages/release/data/experiment/html/minfiData.html" class="external-link">minfiData</a> package does not include genotype data, so we’ll generate some fake data, where we swap the genotypes of two samples:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The minfidata package does not contain genetic data, so here we'll generate some fake data.</span></span>
<span><span class="va">rgset</span> <span class="op">&lt;-</span> <span class="fu">minfi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/minfi/man/read.metharray.html" class="external-link">read.metharray</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Basename</span>, extended<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu">minfi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/minfi/man/RGChannelSet-class.html" class="external-link">getSnpBeta</a></span><span class="op">(</span><span class="va">rgset</span><span class="op">)</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu">omicsPrint</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/omicsPrint/man/beta2genotype.html" class="external-link">beta2genotype</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="va">geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">geno</span> <span class="op">==</span> <span class="fl">3</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">geno</span> <span class="op">==</span> <span class="fl">2</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">geno</span> <span class="op">==</span> <span class="fl">1</span>, <span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Swap a sample </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Run getQCmetrics function with default parameters </span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, chunk_size <span class="op">=</span> <span class="fl">3</span>, geno <span class="op">=</span> <span class="va">geno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the outliers shows that now two samples fail because they do not match their genotype data</span></span>
<span><span class="fu"><a href="../reference/getSampleOutliers.html">getSampleOutliers</a></span><span class="op">(</span><span class="va">qcmetrics</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">IBS_outlier</span><span class="op">)</span> </span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 12</span></span></span>
<span><span class="co">#&gt;   Sample_Name MU_outlier RG_ratio_outlier GR_ratio_outlier OP_outlier HC_outlier</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 5723646052… FALSE      FALSE            FALSE            FALSE      FALSE     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 5723646053… FALSE      FALSE            FALSE            FALSE      FALSE     </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6 more variables: bscon_outlier &lt;lgl&gt;, detectionP_outlier &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   beadNr_outlier &lt;lgl&gt;, sex_outlier &lt;lgl&gt;, IBS_outlier &lt;lgl&gt;, Total &lt;lgl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="infer-snps">Infer SNPs<a class="anchor" aria-label="anchor" href="#infer-snps"></a>
</h3>
<p>In addition to the designated SNP probes, a considerable number of probes overlap with SNPs and may therefore measure genetic variation. These probes can be used to obtain more accurate estimations of relatedness and genotype concordance, as described in the <a href="https://bioconductor.org/packages/release/bioc/html/omicsPrint.html" class="external-link">omicsPrint</a> package <span class="citation">(Iterson et al. 2018)</span>. We have included a set of 426 probes that accurately measure genetic variation in Project MinE data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">QCpG</span><span class="fu">::</span><span class="va"><a href="../reference/snplinker_mine.html">snplinker_mine</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   probe      ID_CPRA         </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> cg12213037 chr1:1666809:G:A</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> cg19405842 chr1:2038662:G:A</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> cg01296877 chr1:2274956:G:A</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> cg10091792 chr1:2844357:G:A</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> cg13928473 chr1:6063655:G:A</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> cg21294301 chr1:8120055:C:T</span></span></code></pre></div>
<p>Include these probes in the <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics_nondesignatedsnps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                                             chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                                             snplinker <span class="op">=</span> <span class="fu">QCpG</span><span class="fu">::</span><span class="va"><a href="../reference/snplinker_mine.html">snplinker_mine</a></span><span class="op">)</span></span></code></pre></div>
<p>Note that Project MinE mainly includes mostly individuals with an European ancestry, for other ancestries other probes may work better.</p>
</div>
<div class="section level3">
<h3 id="reduce-size-of-relationship-matrix">Reduce size of relationship matrix<a class="anchor" aria-label="anchor" href="#reduce-size-of-relationship-matrix"></a>
</h3>
<p>The relatedness output contains N^2 rows (where N is the number of samples), and thus it will rapidly increase in size with large sample sizes. To prevent <code>QCmetrics</code> objects becoming very large, the <code>save_all_relationships</code> parameter can be set to <code>FALSE</code>. In this case only the data for related samples will be saved. Note that the plot showing the relatedness among all samples can still be shown in the <code><a href="../reference/QCapp.html">QCapp()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                          save_all_relationships <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          snplinker <span class="op">=</span> <span class="fu">QCpG</span><span class="fu">::</span><span class="va"><a href="../reference/snplinker_mine.html">snplinker_mine</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="principal-components">Principal components<a class="anchor" aria-label="anchor" href="#principal-components"></a>
</h2>
<p>By default, array-wide PCA is not performed (<code>calculate_PCs = FALSE</code>), since this significantly increases the amount of working memory required. If you have enough RAM available, we recommend setting <code>calculate_PCs = TRUE</code>. RAM usage can be reduced by specifying the <code>n_probes_PCA</code> which specifies the number of most variable probes to perform PCA on. Control probe PCA is always performed, since it only requires a small number of probes. Usually, control probe PCA gives a good indication of technical variation or outliers.</p>
<p>Perform array-wide PCA using the top 50,000 most variable probes:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate PCs, using the top 50,000 most variable probes</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                           n_probes_PCA <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span></span></code></pre></div>
<p>Both control PCs and array-wide PCs will be tested for association with available variables, which are visualized in the <code><a href="../reference/QCapp.html">QCapp()</a></code>. By default the variables tested are: cell type fractions, predicted age, epigenetic smoking score, epigenetic alcohol score and epigenetic BMI score. Additional variables to test can be specified with the <code>heatmap_variables</code> parameter.</p>
<p>Add some technical variables:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The 'heatmap_variables_fct' argument specifies which variables should be treated as a factor</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                           heatmap_variables <span class="op">=</span> <span class="st">"Slide"</span>, </span>
<span>                           heatmap_variables_fct <span class="op">=</span> <span class="st">"Slide"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="qc-settings-and-thresholds">QC settings and thresholds<a class="anchor" aria-label="anchor" href="#qc-settings-and-thresholds"></a>
</h2>
<p>The <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function used a predefined set of thresholds to define outliers. (based on our experience with Project MinE data). These thresholds can be interactively explored and modified within the <code><a href="../reference/QCapp.html">QCapp()</a></code>, however, users can also specify custom thresholds when creating the QCmetrics (note that not all thresholds have to be specified, tunspecified thresholds will be filled with the default thresholds):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set a stricter MU thresholds and a less strict bscon threshold:</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           sample_qc_thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>MU <span class="op">=</span> <span class="fl">2500</span>, bscon <span class="op">=</span> <span class="fl">70</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Two thresholds cannot be modified within the <code><a href="../reference/QCapp.html">QCapp()</a></code>: <code>detp_threshold</code> and <code>beadnr_threshold</code>. These thresholds define the measurements which are considered low quality, and subsequently the percentage of bad quality probes per sample and the percentage of bad quality samples per probe are calculated. These thresholds cannot be adjusted within the app because it would require access to the entire dataset, which is not feasible within the app.</p>
<p>These thresholds can be adjusted when creating the <code>QCmetrics</code> however:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set a stricter beadnr threshold and a less strict detp threshold:</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           detp_threshold <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                           beadnr_threshold <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p>Another option that can be only set when creating the QCmetrics object is the <code>probe_qc_method</code>. This parameter controls whether probe QC should be performed <em>after</em> or <em>before</em> removing samples that fail based on sample QC. By default, probe QC is performed <em>after</em> removing sample outliers. This can be changed as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform probe QC before removing sample outliers</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           probe_qc_method <span class="op">=</span> <span class="st">"pre_sampleqc"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combining-qcmetrics-objects">Combining QCmetrics objects<a class="anchor" aria-label="anchor" href="#combining-qcmetrics-objects"></a>
</h2>
<p><code>QCmetric</code> objects can be combined using the <code><a href="../reference/combineQCmetrics.html">combineQCmetrics()</a></code> method. Note that array-wide PCs (if present) and relatedness data are dropped when combined. Control probe PCs are recalculated, and are thus included in the combined object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>, </span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qcmetrics1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, </span>
<span>                            chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qcmetrics2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">[</span><span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,<span class="op">]</span>, </span>
<span>                            chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qcmetrics_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineQCmetrics.html">combineQCmetrics</a></span><span class="op">(</span><span class="va">qcmetrics1</span>, </span>
<span>                                       <span class="va">qcmetrics2</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"combined QCmetrics"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test if running all data together or running separately and then combine result in same object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">qcmetrics_combined</span><span class="op">@</span><span class="va">Sample_Metrics</span>, <span class="va">qcmetrics</span><span class="op">@</span><span class="va">Sample_Metrics</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Note that <code>combineQCmetrics</code> performs several checks to make sure that the <code>QCmetrics</code> objects can be combined. These include (among others), a check whether the same thresholds are used in the objects and whether they were created using the same package version. For example:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, </span>
<span>                            chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>,probe_qc_method <span class="op">=</span> <span class="st">"pre_sampleqc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Creating a QCmetrics object</span></span>
<span><span class="co">#&gt; Analysis started at: 2024-07-18 11:20:07.89839</span></span>
<span><span class="co">#&gt; Array: 450k</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; Extracting QCmetrics in chunks of 3 samples..</span></span>
<span><span class="co">#&gt; Using 144/144 probes for alcohol prediction</span></span>
<span><span class="co">#&gt; Using 187/187 probes for smoking prediction</span></span>
<span><span class="co">#&gt; Using 435/435 probes for BMI prediction</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; Merging the chunks..</span></span>
<span><span class="co">#&gt; Performing IBS..</span></span>
<span><span class="co">#&gt; 0 out of 65 probes could be accurately called</span></span>
<span><span class="co">#&gt; Calculating relatedness using the DNAm-inferred SNPs</span></span>
<span><span class="co">#&gt; Performing array-wide PCA..</span></span>
<span><span class="co">#&gt; Performing control probe PCA..</span></span>
<span><span class="co">#&gt; Performing association tests between array-wide PCs and technical/biological variables..</span></span>
<span><span class="co">#&gt; The following variables have zero variance, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  CD8T</span></span>
<span><span class="co">#&gt;  Neutro</span></span>
<span><span class="co">#&gt; Performing association tests between control probe PCs and technical/biological variables..</span></span>
<span><span class="co">#&gt; The following variables have zero variance, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  CD8T</span></span>
<span><span class="co">#&gt;  Neutro</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Merging the metrics..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of outliers using default thresholds:</span></span>
<span><span class="co">#&gt; MU: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; RG_ratio: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; GR_ratio: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; OP: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; HC: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; bscon: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; detectionP: 1 outliers (33.33%) [WARNING] High percentage of outliers!</span></span>
<span><span class="co">#&gt; beadNr: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; sex: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; Total: 1 outliers (33.33%) [WARNING] High percentage of outliers!</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; Analysis finished at: 2024-07-18 11:20:33.566935</span></span>
<span></span>
<span><span class="va">qcmetrics2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">[</span><span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,<span class="op">]</span>, </span>
<span>                            chunk_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Creating a QCmetrics object</span></span>
<span><span class="co">#&gt; Analysis started at: 2024-07-18 11:20:33.56821</span></span>
<span><span class="co">#&gt; Array: 450k</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt; Extracting QCmetrics in chunks of 3 samples..</span></span>
<span><span class="co">#&gt; Using 144/144 probes for alcohol prediction</span></span>
<span><span class="co">#&gt; Using 187/187 probes for smoking prediction</span></span>
<span><span class="co">#&gt; Using 435/435 probes for BMI prediction</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; Merging the chunks..</span></span>
<span><span class="co">#&gt; Performing IBS..</span></span>
<span><span class="co">#&gt; 0 out of 65 probes could be accurately called</span></span>
<span><span class="co">#&gt; Calculating relatedness using the DNAm-inferred SNPs</span></span>
<span><span class="co">#&gt; Performing array-wide PCA..</span></span>
<span><span class="co">#&gt; Performing control probe PCA..</span></span>
<span><span class="co">#&gt; Performing association tests between array-wide PCs and technical/biological variables..</span></span>
<span><span class="co">#&gt; The following variables have only one level, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  Sex</span></span>
<span><span class="co">#&gt; The following variables have zero variance, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  CD8T</span></span>
<span><span class="co">#&gt;  Neutro</span></span>
<span><span class="co">#&gt; Performing association tests between control probe PCs and technical/biological variables..</span></span>
<span><span class="co">#&gt; The following variables have only one level, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  Sex</span></span>
<span><span class="co">#&gt; The following variables have zero variance, and will therefore be excluded from the association tests:</span></span>
<span><span class="co">#&gt;  CD8T</span></span>
<span><span class="co">#&gt;  Neutro</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Merging the metrics..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of outliers using default thresholds:</span></span>
<span><span class="co">#&gt; MU: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; RG_ratio: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; GR_ratio: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; OP: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; HC: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; bscon: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; detectionP: 1 outliers (33.33%) [WARNING] High percentage of outliers!</span></span>
<span><span class="co">#&gt; beadNr: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; sex: 0 outliers (0%)</span></span>
<span><span class="co">#&gt; Total: 1 outliers (33.33%) [WARNING] High percentage of outliers!</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; Analysis finished at: 2024-07-18 11:20:59.02073</span></span>
<span></span>
<span><span class="va">qcmetrics_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineQCmetrics.html">combineQCmetrics</a></span><span class="op">(</span><span class="va">qcmetrics1</span>, </span>
<span>                                       <span class="va">qcmetrics2</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"combined QCmetrics"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Array type: 450k</span></span>
<span><span class="co">#&gt; [Check] Same package version was used for all objects</span></span>
<span><span class="co">#&gt; Checking whether thresholds are identical across objects..</span></span>
<span><span class="co">#&gt; [Check] Detection P-value thresholds are identical</span></span>
<span><span class="co">#&gt; [Check] Beadnr thresholds are identical</span></span>
<span><span class="co">#&gt; Error in combine_metrics(x, y, ..., force = force, include_probe_metrics = include_probe_metrics, : Probe qc methods differ between metric-files!</span></span></code></pre></div>
<p>If you do want to proceed combining the objects the <code>force</code> parameter can be set to TRUE:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineQCmetrics.html">combineQCmetrics</a></span><span class="op">(</span><span class="va">qcmetrics1</span>, </span>
<span>                                       <span class="va">qcmetrics2</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"combined QCmetrics"</span>,</span>
<span>                                       force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parallelization">Parallelization<a class="anchor" aria-label="anchor" href="#parallelization"></a>
</h2>
<p>The <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function can be parallelized by specifying the <code>cores</code> parameter, in which case the BiocConductor <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a> package is used to distribute tasks.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check number of cores available:</span></span>
<span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 80</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run in parallel and check run time</span></span>
<span><span class="va">time1_parallel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">qcmetrics_parallel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span>                                    chunk_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                    cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">time2_parallel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">time2_parallel</span><span class="op">-</span><span class="va">time1_parallel</span></span>
<span><span class="co">#&gt; Time difference of 1.021751 mins</span></span>
<span></span>
<span><span class="co"># Run single-threaded and check run time</span></span>
<span><span class="va">time1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">time2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">time2</span> <span class="op">-</span>  <span class="va">time1</span></span>
<span><span class="co">#&gt; Time difference of 1.006328 mins</span></span></code></pre></div>
<p>By default, the <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">BiocParallel::MulticoreParam()</a></code> backend is used, which works well on non-Windows computers. On Windows computers or cluster computers other backends (such as <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">BiocParallel::SnowParam()</a></code>) can be specified using the <code>BPPARAM</code> parameter, see for more information.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use Snow backend</span></span>
<span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span>                           chunk_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           cores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"SOCK"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="masking-low-quality-measurements">Masking low-quality measurements<a class="anchor" aria-label="anchor" href="#masking-low-quality-measurements"></a>
</h2>
<p>The <code>path_failed_measurements</code> parameter in <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> allows for saving a sparse matrix that indicates for each measurement whether it is of low quality (based on the detp_threshold (default=1e-16) and beadnr_threshold (default=3)). This matrix can bed used to mask low-quality measurements in the beta-matrix using the <code><a href="../reference/mask_measurements.html">mask_measurements()</a></code> function.</p>
<p>First, we’ll generate QCmetrics while saving the matrix indicating low-quality measurements:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qcmetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getQCmetrics.html">getQCmetrics</a></span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span>                           calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           path_failed_measurements <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/mask_measurements.rds"</span>, <span class="va">tempdir</span><span class="op">)</span></span>
<span>                           <span class="op">)</span></span>
<span><span class="co">#&gt; Warning in Matrix.DeprecatedCoerce(cd1, cd2): 'as(&lt;dgCMatrix&gt;, "lgCMatrix")' is deprecated.</span></span>
<span><span class="co">#&gt; Use 'as(., "lMatrix")' instead.</span></span>
<span><span class="co">#&gt; See help("Deprecated") and help("Matrix-deprecated").</span></span></code></pre></div>
<p>Save example betas to mask measurements in:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rgset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/minfi/man/read.metharray.html" class="external-link">read.metharray</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Basename</span>, extended<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/minfi/man/MethylSet-class.html" class="external-link">getBeta</a></span><span class="op">(</span><span class="va">rgset</span><span class="op">)</span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>Probe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">beta</span><span class="op">)</span>, <span class="va">beta</span><span class="op">)</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/beta.txt"</span>, <span class="va">tempdir</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Mask measurements using the <code><a href="../reference/mask_measurements.html">mask_measurements()</a></code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/mask_measurements.rds"</span>, <span class="va">tempdir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/mask_measurements.html">mask_measurements</a></span><span class="op">(</span>beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/beta.txt"</span>, <span class="va">tempdir</span><span class="op">)</span>, output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/beta_masked.txt"</span>, <span class="va">tempdir</span><span class="op">)</span>, mask_matrix <span class="op">=</span> <span class="va">mask_matrix</span> <span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; cols(</span></span>
<span><span class="co">#&gt;   Probe = <span style="color: #BB0000;">col_character()</span>,</span></span>
<span><span class="co">#&gt;   `5723646052_R02C02` = <span style="color: #00BB00;">col_double()</span>,</span></span>
<span><span class="co">#&gt;   `5723646052_R04C01` = <span style="color: #00BB00;">col_double()</span>,</span></span>
<span><span class="co">#&gt;   `5723646052_R05C02` = <span style="color: #00BB00;">col_double()</span>,</span></span>
<span><span class="co">#&gt;   `5723646053_R04C02` = <span style="color: #00BB00;">col_double()</span>,</span></span>
<span><span class="co">#&gt;   `5723646053_R05C02` = <span style="color: #00BB00;">col_double()</span>,</span></span>
<span><span class="co">#&gt;   `5723646053_R06C02` = <span style="color: #00BB00;">col_double()</span></span></span>
<span><span class="co">#&gt; )</span></span>
<span><span class="co">#&gt; |=                                                                    | 2% 1 MB|=                                                                    | 2% 1 MB|==                                                                   | 2% 1 MB|==                                                                   | 3% 1 MB|==                                                                   | 3% 1 MB|==                                                                   | 3% 2 MB|==                                                                   | 4% 2 MB|==                                                                   | 4% 2 MB|===                                                                  | 4% 2 MB|===                                                                  | 4% 2 MB|===                                                                  | 5% 2 MB|===                                                                  | 5% 2 MB|===                                                                  | 5% 3 MB|====                                                                 | 5% 3 MB|====                                                                 | 6% 3 MB|====                                                                 | 6% 3 MB|====                                                                 | 6% 3 MB|====                                                                 | 6% 3 MB|====                                                                 | 7% 3 MB|=====                                                                | 7% 4 MB|=====                                                                | 7% 4 MB|=====                                                                | 8% 4 MB|=====                                                                | 8% 4 MB|=====                                                                | 8% 4 MB|======                                                               | 8% 4 MB|======                                                               | 8% 4 MB|======                                                               | 9% 5 MB|======                                                               | 9% 5 MB|======                                                               | 9% 5 MB|======                                                              | 10% 5 MB|=======                                                             | 10% 5 MB|=======                                                             | 10% 5 MB|=======                                                             | 10% 5 MB|=======                                                             | 10% 6 MB|=======                                                             | 11% 6 MB|=======                                                             | 11% 6 MB|========                                                            | 11% 6 MB|========                                                            | 12% 6 MB|========                                                            | 12% 6 MB|========                                                            | 12% 6 MB|========                                                            | 12% 7 MB|=========                                                           | 13% 7 MB|=========                                                           | 13% 7 MB|=========                                                           | 13% 7 MB|=========                                                           | 13% 7 MB|=========                                                           | 14% 7 MB|==========                                                          | 14% 8 MB|==========                                                          | 14% 8 MB|==========                                                          | 14% 8 MB|==========                                                          | 15% 8 MB|==========                                                          | 15% 8 MB|==========                                                          | 15% 8 MB|===========                                                         | 16% 8 MB|===========                                                         | 16% 9 MB|===========                                                         | 16% 9 MB|===========                                                         | 16% 9 MB|===========                                                         | 16% 9 MB|===========                                                         | 17% 9 MB|============                                                        | 17% 9 MB|============                                                        | 17% 9 MB|============                                                       | 18% 10 MB|============                                                       | 18% 10 MB|============                                                       | 18% 10 MB|============                                                       | 18% 10 MB|============                                                       | 18% 10 MB|=============                                                      | 19% 10 MB|=============                                                      | 19% 10 MB|=============                                                      | 19% 10 MB|=============                                                      | 20% 11 MB|=============                                                      | 20% 11 MB|==============                                                     | 20% 11 MB|==============                                                     | 20% 11 MB|==============                                                     | 21% 11 MB|==============                                                     | 21% 11 MB|==============                                                     | 21% 11 MB|==============                                                     | 21% 12 MB|===============                                                    | 22% 12 MB|===============                                                    | 22% 12 MB|===============                                                    | 22% 12 MB|===============                                                    | 22% 12 MB|===============                                                    | 23% 12 MB|===============                                                    | 23% 12 MB|================                                                   | 23% 13 MB|================                                                   | 24% 13 MB|================                                                   | 24% 13 MB|================                                                   | 24% 13 MB|================                                                   | 24% 13 MB|================                                                   | 24% 13 MB|=================                                                  | 25% 13 MB|=================                                                  | 25% 14 MB|=================                                                  | 25% 14 MB|=================                                                  | 26% 14 MB|=================                                                  | 26% 14 MB|==================                                                 | 26% 14 MB|==================                                                 | 27% 14 MB|==================                                                 | 27% 14 MB|==================                                                 | 27% 15 MB|==================                                                 | 27% 15 MB|==================                                                 | 27% 15 MB|===================                                                | 28% 15 MB|===================                                                | 28% 15 MB|===================                                                | 28% 15 MB|===================                                                | 29% 16 MB|===================                                                | 29% 16 MB|===================                                                | 29% 16 MB|====================                                               | 29% 16 MB|====================                                               | 29% 16 MB|====================                                               | 30% 16 MB|====================                                               | 30% 16 MB|====================                                               | 30% 17 MB|=====================                                              | 31% 17 MB|=====================                                              | 31% 17 MB|=====================                                              | 31% 17 MB|=====================                                              | 31% 17 MB|=====================                                              | 32% 17 MB|=====================                                              | 32% 17 MB|======================                                             | 32% 18 MB|======================                                             | 32% 18 MB|======================                                             | 33% 18 MB|======================                                             | 33% 18 MB|======================                                             | 33% 18 MB|======================                                             | 33% 18 MB|=======================                                            | 34% 18 MB|=======================                                            | 34% 19 MB|=======================                                            | 34% 19 MB|=======================                                            | 34% 19 MB|=======================                                            | 35% 19 MB|=======================                                            | 35% 19 MB|========================                                           | 35% 19 MB|========================                                           | 35% 19 MB|========================                                           | 36% 19 MB|========================                                           | 36% 20 MB|========================                                           | 36% 20 MB|=========================                                          | 37% 20 MB|=========================                                          | 37% 20 MB|=========================                                          | 37% 20 MB|=========================                                          | 37% 20 MB|=========================                                          | 37% 20 MB|=========================                                          | 38% 21 MB|==========================                                         | 38% 21 MB|==========================                                         | 38% 21 MB|==========================                                         | 39% 21 MB|==========================                                         | 39% 21 MB|==========================                                         | 39% 21 MB|==========================                                         | 39% 21 MB|===========================                                        | 39% 22 MB|===========================                                        | 40% 22 MB|===========================                                        | 40% 22 MB|===========================                                        | 40% 22 MB|===========================                                        | 41% 22 MB|============================                                       | 41% 22 MB|============================                                       | 41% 22 MB|============================                                       | 41% 23 MB|============================                                       | 41% 23 MB|============================                                       | 42% 23 MB|============================                                       | 42% 23 MB|=============================                                      | 42% 23 MB|=============================                                      | 43% 23 MB|=============================                                      | 43% 24 MB|=============================                                      | 43% 24 MB|=============================                                      | 43% 24 MB|=============================                                      | 44% 24 MB|==============================                                     | 44% 24 MB|==============================                                     | 44% 24 MB|==============================                                     | 44% 24 MB|==============================                                     | 45% 25 MB|==============================                                     | 45% 25 MB|==============================                                     | 45% 25 MB|===============================                                    | 45% 25 MB|===============================                                    | 46% 25 MB|===============================                                    | 46% 25 MB|===============================                                    | 46% 25 MB|===============================                                    | 46% 25 MB|================================                                   | 47% 26 MB|================================                                   | 47% 26 MB|================================                                   | 47% 26 MB|================================                                   | 47% 26 MB|================================                                   | 48% 26 MB|================================                                   | 48% 26 MB|=================================                                  | 48% 26 MB|=================================                                  | 49% 27 MB|=================================                                  | 49% 27 MB|=================================                                  | 49% 27 MB|=================================                                  | 49% 27 MB|=================================                                  | 49% 27 MB|==================================                                 | 50% 27 MB|==================================                                 | 50% 27 MB|==================================                                 | 50% 28 MB|==================================                                 | 51% 28 MB|==================================                                 | 51% 28 MB|===================================                                | 51% 28 MB|===================================                                | 51% 28 MB|===================================                                | 51% 28 MB|===================================                                | 52% 28 MB|===================================                                | 52% 29 MB|===================================                                | 52% 29 MB|====================================                               | 53% 29 MB|====================================                               | 53% 29 MB|====================================                               | 53% 29 MB|====================================                               | 53% 29 MB|====================================                               | 54% 29 MB|====================================                               | 54% 30 MB|=====================================                              | 54% 30 MB|=====================================                              | 54% 30 MB|=====================================                              | 55% 30 MB|=====================================                              | 55% 30 MB|=====================================                              | 55% 30 MB|=====================================                              | 55% 30 MB|======================================                             | 56% 31 MB|======================================                             | 56% 31 MB|======================================                             | 56% 31 MB|======================================                             | 56% 31 MB|======================================                             | 57% 31 MB|=======================================                            | 57% 31 MB|=======================================                            | 57% 31 MB|=======================================                            | 57% 31 MB|=======================================                            | 58% 32 MB|=======================================                            | 58% 32 MB|=======================================                            | 58% 32 MB|========================================                           | 58% 32 MB|========================================                           | 59% 32 MB|========================================                           | 59% 32 MB|========================================                           | 59% 33 MB|========================================                           | 59% 33 MB|========================================                           | 60% 33 MB|=========================================                          | 60% 33 MB|=========================================                          | 60% 33 MB|=========================================                          | 61% 33 MB|=========================================                          | 61% 33 MB|=========================================                          | 61% 34 MB|==========================================                         | 61% 34 MB|==========================================                         | 61% 34 MB|==========================================                         | 62% 34 MB|==========================================                         | 62% 34 MB|==========================================                         | 62% 34 MB|==========================================                         | 63% 34 MB|===========================================                        | 63% 35 MB|===========================================                        | 63% 35 MB|===========================================                        | 63% 35 MB|===========================================                        | 63% 35 MB|===========================================                        | 64% 35 MB|===========================================                        | 64% 35 MB|============================================                       | 64% 35 MB|============================================                       | 65% 36 MB|============================================                       | 65% 36 MB|============================================                       | 65% 36 MB|============================================                       | 66% 36 MB|============================================                       | 66% 36 MB|=============================================                      | 66% 36 MB|=============================================                      | 66% 36 MB|=============================================                      | 66% 36 MB|=============================================                      | 67% 37 MB|=============================================                      | 67% 37 MB|==============================================                     | 67% 37 MB|==============================================                     | 68% 37 MB|==============================================                     | 68% 37 MB|==============================================                     | 68% 37 MB|==============================================                     | 68% 37 MB|==============================================                     | 68% 38 MB|===============================================                    | 69% 38 MB|===============================================                    | 69% 38 MB|===============================================                    | 69% 38 MB|===============================================                    | 70% 38 MB|===============================================                    | 70% 38 MB|===============================================                    | 70% 38 MB|================================================                   | 70% 39 MB|================================================                   | 70% 39 MB|================================================                   | 71% 39 MB|================================================                   | 71% 39 MB|================================================                   | 71% 39 MB|=================================================                  | 72% 39 MB|=================================================                  | 72% 39 MB|=================================================                  | 72% 40 MB|=================================================                  | 72% 40 MB|=================================================                  | 73% 40 MB|=================================================                  | 73% 40 MB|==================================================                 | 73% 40 MB|==================================================                 | 73% 40 MB|==================================================                 | 74% 41 MB|==================================================                 | 74% 41 MB|==================================================                 | 74% 41 MB|==================================================                 | 74% 41 MB|===================================================                | 75% 41 MB|===================================================                | 75% 41 MB|===================================================                | 75% 41 MB|===================================================                | 75% 42 MB|===================================================                | 76% 42 MB|===================================================                | 76% 42 MB|====================================================               | 76% 42 MB|====================================================               | 76% 42 MB|====================================================               | 77% 42 MB|====================================================               | 77% 42 MB|====================================================               | 77% 42 MB|=====================================================              | 78% 43 MB|=====================================================              | 78% 43 MB|=====================================================              | 78% 43 MB|=====================================================              | 78% 43 MB|=====================================================              | 78% 43 MB|=====================================================              | 79% 43 MB|======================================================             | 79% 43 MB|======================================================             | 79% 44 MB|======================================================             | 80% 44 MB|======================================================             | 80% 44 MB|======================================================             | 80% 44 MB|======================================================             | 80% 44 MB|=======================================================            | 80% 44 MB|=======================================================            | 81% 44 MB|=======================================================            | 81% 45 MB|=======================================================            | 81% 45 MB|=======================================================            | 82% 45 MB|========================================================           | 82% 45 MB|========================================================           | 82% 45 MB|========================================================           | 82% 45 MB|========================================================           | 83% 45 MB|========================================================           | 83% 46 MB|========================================================           | 83% 46 MB|=========================================================          | 83% 46 MB|=========================================================          | 84% 46 MB|=========================================================          | 84% 46 MB|=========================================================          | 84% 46 MB|=========================================================          | 84% 46 MB|=========================================================          | 85% 47 MB|==========================================================         | 85% 47 MB|==========================================================         | 85% 47 MB|==========================================================         | 85% 47 MB|==========================================================         | 86% 47 MB|==========================================================         | 86% 47 MB|==========================================================         | 86% 47 MB|===========================================================        | 86% 48 MB|===========================================================        | 87% 48 MB|===========================================================        | 87% 48 MB|===========================================================        | 87% 48 MB|===========================================================        | 88% 48 MB|============================================================       | 88% 48 MB|============================================================       | 88% 48 MB|============================================================       | 88% 48 MB|============================================================       | 88% 49 MB|============================================================       | 89% 49 MB|============================================================       | 89% 49 MB|=============================================================      | 89% 49 MB|=============================================================      | 90% 49 MB|=============================================================      | 90% 49 MB|=============================================================      | 90% 50 MB|=============================================================      | 90% 50 MB|=============================================================      | 90% 50 MB|==============================================================     | 91% 50 MB|==============================================================     | 91% 50 MB|==============================================================     | 91% 50 MB|==============================================================     | 92% 50 MB|==============================================================     | 92% 51 MB|===============================================================    | 92% 51 MB|===============================================================    | 92% 51 MB|===============================================================    | 92% 51 MB|===============================================================    | 93% 51 MB|===============================================================    | 93% 51 MB|===============================================================    | 93% 51 MB|================================================================   | 94% 52 MB|================================================================   | 94% 52 MB|================================================================   | 94% 52 MB|================================================================   | 94% 52 MB|================================================================   | 95% 52 MB|================================================================   | 95% 52 MB|=================================================================  | 95% 52 MB|=================================================================  | 95% 53 MB|=================================================================  | 96% 53 MB|=================================================================  | 96% 53 MB|=================================================================  | 96% 53 MB|=================================================================  | 96% 53 MB|================================================================== | 97% 53 MB|================================================================== | 97% 53 MB|================================================================== | 97% 54 MB|================================================================== | 97% 54 MB|================================================================== | 98% 54 MB|===================================================================| 98% 54 MB|===================================================================| 98% 54 MB|===================================================================| 98% 54 MB|===================================================================| 99% 54 MB|===================================================================| 99% 55 MB|===================================================================| 99% 55 MB|===================================================================| 100% 55 MB</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="va">beta_masked</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s/beta_masked.txt"</span>, <span class="va">tempdir</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">485512</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">7</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (1): Probe</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (6): 5723646052_R02C02, 5723646052_R04C01, 5723646052_R05C02, 5723646053...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">beta_masked</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 171329</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-van_iterson_omicsprint_2018">
<p>Iterson, Maarten van, Davy Cats, Paul Hop, BIOS Consortium, and Bastiaan T Heijmans. 2018. “omicsPrint: Detection of Data Linkage Errors in Multiple Omics Studies.” Edited by Oliver Stegle. <em>Bioinformatics</em> 34 (12): 2142–3. <a href="https://doi.org/10.1093/bioinformatics/bty062" class="external-link">https://doi.org/10.1093/bioinformatics/bty062</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
