<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Available metrics • QCpG</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Available metrics">
<meta property="og:description" content="A description of the metrics available in the sample QC as well as the probe QC.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QCpG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AdvancedOptions.html">Advanced Options</a>
    </li>
    <li>
      <a href="../articles/AvailableMetrics.html">Available metrics</a>
    </li>
    <li>
      <a href="../articles/EWASworkflow.html">Example EWAS workflow</a>
    </li>
    <li>
      <a href="../articles/GetExampleData.html">Get example data</a>
    </li>
    <li>
      <a href="../articles/GettingStarted.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/UsingTheApp.html">Using the QC app</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="AvailableMetrics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Available metrics</h1>
            
      

      <div class="hidden name"><code>AvailableMetrics.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="sample-qc">Sample QC<a class="anchor" aria-label="anchor" href="#sample-qc"></a>
</h2>
<ul>
<li>
<p><strong>PCA</strong></p>
<ul>
<li>
<em>Control probe PCA</em>: PCA performed on the various control probes that are present on the array, thereby measuring and summarizing various sources of technical variation.</li>
<li>
<em>Array-wide PCA</em>: PCA performed on all probes on the array. By default, array-wide PCA is not performed (<code>calculate_PCs = FALSE</code>), since this significantly increases the amount of working memory required. If you have enough RAM available, we recommend setting <code>calculate_PCs = TRUE</code>. RAM usage can be reduced by specifying the <code>n_probes_PCA</code> which specifies the number of most variable probes to perform PCA on. Control probe PCA is always performed, since it only requires a small number of probes. Usually, control probe PCA gives a good indication of technical variation or outliers.</li>
</ul>
</li>
<li><p><strong>Sex</strong>: Median X and Y copy number, as implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/minfi.html" class="external-link">minfi</a> package <span class="citation">(Aryee et al. 2014)</span>. The lower plot shows the difference (median(Y) - median(X)), by default samples with diff &lt; -2 are classified as females.</p></li>
<li><p><strong>MU</strong>: Median methylated signal vs. median unmethylated signal.</p></li>
<li><p><strong>RG</strong>: Median green vs median red signal. Note that these intensities are calculated in type I probes. Since both unmethylated and methylated signals are measured using the same color in type I probes, the sum of the M and U signals in these probes should be invariant with respect to actual methylation levels (in contrast to the MU plot).</p></li>
<li><p><strong>OP</strong>: Rotated OP plot based on non-polymorphic control probes as implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/MethylAid.html" class="external-link">MethylAid</a> package <span class="citation">(Iterson et al. 2014)</span>.</p></li>
<li><p><strong>HC</strong>: Rotated HC plot based on hybridization control probes (sample-independent as implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/MethylAid.html" class="external-link">MethylAid</a> package <span class="citation">(Iterson et al. 2014)</span>.</p></li>
<li><p><strong>bscon</strong>: bscon metric from the <a href="https://www.bioconductor.org/packages/release/bioc/html/wateRmelon.html" class="external-link">wateRmelon</a> package, which measures bisulfite conversion efficiency <span class="citation">(Pidsley et al. 2013)</span>.</p></li>
<li><p><strong>detP</strong>: Percentage of probes with detection P-value &gt; threshold. This threshold can be set using the <code>detP_threshold</code> parameter in the <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function.</p></li>
<li><p><strong>beadNr</strong>: Percentage of probes with beadNr &lt; threshold. This threshold can be set using the <code>beadNr_threshold</code> parameter in the <code><a href="../reference/getQCmetrics.html">getQCmetrics()</a></code> function.</p></li>
<li><p><strong>IBS</strong>: For each sample, DNAm-derived SNPs are compared with genotype array/WGS SNPs by IBS (identity by state), as implemented in the <a href="https://www.bioconductor.org/packages/release/bioc/html/omicsPrint.html" class="external-link">omicsPrint</a> package. Note that this tab will be blank if no genotype data is supplied <span class="citation">(Iterson et al. 2018)</span>.</p></li>
<li><p><strong>Custom</strong>: Use columns included in the samplesheet as additional QC metrics (should be coded as TRUE/FALSE)</p></li>
</ul>
<p><strong>PMS</strong>: Contains predicted phenotypes (poly-methylation scores): three age predictors <span class="citation">(<span class="citeproc-not-found" data-reference-id="horvath_dna_2013m"><strong>???</strong></span>)</span>, smoking <span class="citation">(Bollepalli et al. 2019)</span>, alcohol <span class="citation">(Liu et al. 2018)</span>, BMI <span class="citation">(Hamilton et al. 2019)</span>, and white blood cell fractions <span class="citation">(Zheng et al. 2018)</span>.</p>
</div>
<div class="section level2">
<h2 id="probe-qc">Probe QC<a class="anchor" aria-label="anchor" href="#probe-qc"></a>
</h2>
<ul>
<li>
<strong>detP</strong>: Percentage of samples with detection <em>P</em>-value &gt; threshold.</li>
<li>
<strong>beadNr</strong>: Percentage of samples with beadNr &lt;= threshold.</li>
<li>
<strong>Masking</strong>: Define probes that should me masked based on cross-reactivity, SNPs in probes and mapping issues. Currently the annotation by Zhou <em>et al.</em> is included <span class="citation">(Zhou, Laird, and Shen 2017)</span>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-aryee_minfi_2014">
<p>Aryee, Martin J, Andrew E Jaffe, Hector Corrada-Bravo, Christine Ladd-Acosta, Andrew P Feinberg, Kasper D Hansen, and Rafael A Irizarry. 2014. “Minfi: A Flexible and Comprehensive Bioconductor Package for the Analysis of Infinium DNA Methylation Microarrays.” <em>Bioinformatics</em> 30 (10): 1363–9.</p>
</div>
<div id="ref-bollepalli_epismoker_2019">
<p>Bollepalli, Sailalitha, Tellervo Korhonen, Jaakko Kaprio, Simon Anders, and Miina Ollikainen. 2019. “EpiSmokEr: A Robust Classifier to Determine Smoking Status from DNA Methylation Data.” <em>Epigenomics</em> 11 (13): 1469–86.</p>
</div>
<div id="ref-hamilton_epigenetic_2019">
<p>Hamilton, Olivia K. L., Qian Zhang, Allan F. McRae, Rosie M. Walker, Stewart W. Morris, Paul Redmond, Archie Campbell, et al. 2019. “An Epigenetic Score for BMI Based on DNA Methylation Correlates with Poor Physical Health and Major Disease in the Lothian Birth Cohort.” <em>International Journal of Obesity</em> 43 (September): 1795–1802. <a href="https://doi.org/10.1038/s41366-018-0262-3" class="external-link">https://doi.org/10.1038/s41366-018-0262-3</a>.</p>
</div>
<div id="ref-van_iterson_omicsprint_2018">
<p>Iterson, Maarten van, Davy Cats, Paul Hop, BIOS Consortium, and Bastiaan T Heijmans. 2018. “omicsPrint: Detection of Data Linkage Errors in Multiple Omics Studies.” Edited by Oliver Stegle. <em>Bioinformatics</em> 34 (12): 2142–3. <a href="https://doi.org/10.1093/bioinformatics/bty062" class="external-link">https://doi.org/10.1093/bioinformatics/bty062</a>.</p>
</div>
<div id="ref-van_iterson_methylaid_2014">
<p>Iterson, Maarten van, Elmar W Tobi, Roderick C Slieker, Wouter den Hollander, René Luijk, P Eline Slagboom, and Bastiaan T Heijmans. 2014. “MethylAid: Visual and Interactive Quality Control of Large Illumina 450k Datasets.” <em>Bioinformatics</em> 30 (23): 3435–7.</p>
</div>
<div id="ref-liu_dna_2018">
<p>Liu, Congyue, Riccardo E. Marioni, ÅA K. Hedman, Liliane Pfeiffer, Pei-Chien Tsai, Lindsay M. Reynolds, Allan C. Just, Qing Duan, Cindy G. Boer, and Toshiko Tanaka. 2018. “A DNA Methylation Biomarker of Alcohol Consumption.” <em>Molecular Psychiatry</em> 23: 422–33.</p>
</div>
<div id="ref-pidsley_data-driven_2013">
<p>Pidsley, Ruth, Chloe CY Wong, Manuela Volta, Katie Lunnon, Jonathan Mill, and Leonard C Schalkwyk. 2013. “A Data-Driven Approach to Preprocessing Illumina 450K Methylation Array Data.” <em>BMC Genomics</em> 14 (293): 1–10.</p>
</div>
<div id="ref-zhang_improved_2019">
<p>Zhang, Qian, Costanza L. Vallerga, Rosie M. Walker, Tian Lin, Anjali K. Henders, Grant W. Montgomery, Ji He, et al. 2019. “Improved Precision of Epigenetic Clock Estimates Across Tissues and Its Implication for Biological Ageing.” <em>Genome Medicine</em> 11 (54): 54. <a href="https://doi.org/10.1186/s13073-019-0667-1" class="external-link">https://doi.org/10.1186/s13073-019-0667-1</a>.</p>
</div>
<div id="ref-zheng_identification_2018">
<p>Zheng, Shijie C., Charles E. Breeze, Stephan Beck, and Andrew E. Teschendorff. 2018. “Identification of Differentially Methylated Cell Types in Epigenome-Wide Association Studies.” <em>Nature Methods</em> 15 (12): 1059–66. <a href="https://doi.org/10.1038/s41592-018-0213-x" class="external-link">https://doi.org/10.1038/s41592-018-0213-x</a>.</p>
</div>
<div id="ref-zhou_comprehensive_2017">
<p>Zhou, Wanding, Peter W. Laird, and Hui Shen. 2017. “Comprehensive Characterization, Annotation and Innovative Use of Infinium DNA Methylation BeadChip Probes.” <em>Nucleic Acids Research</em> 45 (4): e22. <a href="https://doi.org/10.1093/nar/gkw967" class="external-link">https://doi.org/10.1093/nar/gkw967</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
