<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a QC-metrics file — get_qcmetrics • MinEDNAmQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create a QC-metrics file — get_qcmetrics" />
<meta property="og:description" content="Extract QC metrics and outlier information from IDAT files.
Returns summarized QC metrics, sample and probe outliers, relatedness info and
DNA-methylation-based predictors (PMSs) of various traits/exposures
in a QCmetrics object, which can be queried interactively using the QCapp." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MinEDNAmQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/GettingStarted.html">Getting Started</a>
</li>
<li>
  <a href="../articles/UsingTheApp.html">Using the QCapp</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    In-depth articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/GetExampleData.html">Get example data</a>
    </li>
    <li>
      <a href="../articles/AdvancedOptions.html">Advanced Options</a>
    </li>
    <li>
      <a href="../articles/EWASworkflow.html">EWAS workflow</a>
    </li>
    <li>
      <a href="../articles/AvailableMetrics.html">Available Metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/pj_hop">
    <span class="fas fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/pjhop/MinEDNAmQC">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a QC-metrics file</h1>
    
    <div class="hidden name"><code>get_qcmetrics.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract QC metrics and outlier information from IDAT files.
Returns summarized QC metrics, sample and probe outliers, relatedness info and
DNA-methylation-based predictors (PMSs) of various traits/exposures
in a <a href='QCmetrics-class.html'>QCmetrics</a> object, which can be queried interactively using the <code><a href='QCapp.html'>QCapp</a></code>.</p>
    </div>

    <pre class="usage"><span class='fu'>get_qcmetrics</span><span class='op'>(</span>
  <span class='va'>samplesheet</span>,
  name <span class='op'>=</span> <span class='cn'>NULL</span>,
  experimenter <span class='op'>=</span> <span class='cn'>NULL</span>,
  chunk_size <span class='op'>=</span> <span class='fl'>40</span>,
  geno <span class='op'>=</span> <span class='cn'>NULL</span>,
  snplinker <span class='op'>=</span> <span class='cn'>NULL</span>,
  logfile <span class='op'>=</span> <span class='cn'>NULL</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  probe_qc_method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"post_sampleqc"</span>, <span class='st'>"pre_sampleqc"</span><span class='op'>)</span>,
  heatmap_variables <span class='op'>=</span> <span class='cn'>NULL</span>,
  heatmap_variables_fct <span class='op'>=</span> <span class='cn'>NULL</span>,
  sample_qc_thresholds <span class='op'>=</span> <span class='cn'>NULL</span>,
  probe_qc_thresholds <span class='op'>=</span> <span class='cn'>NULL</span>,
  detp_threshold <span class='op'>=</span> <span class='fl'>1e-16</span>,
  beadnr_threshold <span class='op'>=</span> <span class='fl'>3</span>,
  relatedness_thresholds <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>IBS_mean <span class='op'>=</span> <span class='fl'>1.5</span>, IBS_var <span class='op'>=</span> <span class='fl'>0.35</span><span class='op'>)</span>,
  custom_outliers <span class='op'>=</span> <span class='cn'>NULL</span>,
  epidish_method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"RPC"</span>, <span class='st'>"CBS"</span>, <span class='st'>"CP"</span><span class='op'>)</span>,
  tissue <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"blood"</span>, <span class='st'>"breast"</span>, <span class='st'>"epithelium"</span><span class='op'>)</span>,
  calculate_PCs <span class='op'>=</span> <span class='cn'>FALSE</span>,
  n_probes_PCA <span class='op'>=</span> <span class='cn'>NULL</span>,
  n_estimate_PCA <span class='op'>=</span> <span class='va'>chunk_size</span>,
  nrPCs <span class='op'>=</span> <span class='fl'>10</span>,
  save_all_relationships <span class='op'>=</span> <span class='cn'>NULL</span>,
  path_failed_measurements <span class='op'>=</span> <span class='cn'>NULL</span>,
  cores <span class='op'>=</span> <span class='fl'>1</span>,
  BPPARAM <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>samplesheet</th>
      <td><p>Samplesheet. Should contain the following columns: Sample_Name','Sex' (M/F) and 'Basename'.</p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p>Optionally, provide a name for the experiment (will be included in metadata).</p></td>
    </tr>
    <tr>
      <th>experimenter</th>
      <td><p>Optionally, name of the experiment (will be included in metadata).</p></td>
    </tr>
    <tr>
      <th>chunk_size</th>
      <td><p>Size of chunks, default = 40. Larger chunks will be a little bit faster
but will require more RAM.</p></td>
    </tr>
    <tr>
      <th>geno</th>
      <td><p>An optional matrix with genotypes (rows = snps, columns = samples)
to check for concordance between DNAm and genotype data.</p></td>
    </tr>
    <tr>
      <th>snplinker</th>
      <td><p>Two-column data frame that links DNAm probes to genotype probes (first column = DNAm probes, second column = genotype IDs).
If not provided the 65 designated SNP-probes will be used.</p></td>
    </tr>
    <tr>
      <th>logfile</th>
      <td><p>Filepath to write logfile to, by default no logfile is written (`logfile = NULL`).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>If `TRUE` (the default), information is printed in the console.</p></td>
    </tr>
    <tr>
      <th>probe_qc_method</th>
      <td><p>Should probe QC be performed after sample QC ('post_sampleqc') or prior to sample QC ('pre_sampleqc')?.
If 'post_sampleqc', sample outliers will be removed before performing sample QC.</p></td>
    </tr>
    <tr>
      <th>heatmap_variables</th>
      <td><p>Character vector specifying additional variables to test for association with PCs (default = `NULL`).</p></td>
    </tr>
    <tr>
      <th>heatmap_variables_fct</th>
      <td><p>Relevant if the `heatmap_variables` argument is not `NULL`.
Which of the variables specified in `heatmap_variables` should be treated as factors?</p></td>
    </tr>
    <tr>
      <th>sample_qc_thresholds</th>
      <td><p>List of sample QC thresholds, if `NULL` (the default),
default thresholds will be used.
Note that custom thresholds for a subset of metrics can be specified, default
thresholds will be used for metrics that are not specified.</p></td>
    </tr>
    <tr>
      <th>probe_qc_thresholds</th>
      <td><p>List of probe QC thresholds, if `NULL` (the default),
default thresholds will be used.</p></td>
    </tr>
    <tr>
      <th>detp_threshold</th>
      <td><p>detection P-value threshold (default = 1e-16).</p></td>
    </tr>
    <tr>
      <th>beadnr_threshold</th>
      <td><p>beadnr threshold (default = 3).</p></td>
    </tr>
    <tr>
      <th>relatedness_thresholds</th>
      <td><p>relatedness thresholds.</p></td>
    </tr>
    <tr>
      <th>custom_outliers</th>
      <td><p>Optional additional column name(s) in the samplesheet that
that indicate that samples should be treated as outliers.
The indicated columns should be of class `logical`, (`TRUE` = outlier).</p></td>
    </tr>
    <tr>
      <th>epidish_method</th>
      <td><p>Which method to using in the EpiDish <a href='https://rdrr.io/pkg/EpiDISH/man/epidish.html'>epidish</a> cellcount prediction (RPC, CBS, or CP).
Default = 'RPC'.</p></td>
    </tr>
    <tr>
      <th>tissue</th>
      <td><p>Tissue-type for cellcount-estimation.
Currently 'blood', 'breast' and 'epithelium' are implemented, following the <a href='https://rdrr.io/pkg/EpiDISH/man/epidish.html'>epidish</a> function.</p></td>
    </tr>
    <tr>
      <th>calculate_PCs</th>
      <td><p>Should PCA be performed? (TRUE/FALSE).
Default = `TRUE`. For large sample sizes RAM could become a bottleneck.
To limit RAM usage (besides setting this parameter to `FALSE`), the
`n_probes_PCA` argument can be used to limit PCA to a subset of most variable probes.</p></td>
    </tr>
    <tr>
      <th>n_probes_PCA</th>
      <td><p>Number of most variable probes to maintain for PCA,
defaults to 'NULL' (all probes are used).</p></td>
    </tr>
    <tr>
      <th>n_estimate_PCA</th>
      <td><p>Relevant if `n_probes_PCA` is not `NULL`.
This argument specifies the number of samples used to estimate the most variable probes (defaults to 'chunk_size').</p></td>
    </tr>
    <tr>
      <th>nrPCs</th>
      <td><p>nr of PCs to retain. Default = 10</p></td>
    </tr>
    <tr>
      <th>save_all_relationships</th>
      <td><p>Save all relatedness data (IBS)?
Relatedness output contains N^2 rows, and thus it will rapidly increase in size with large sample sizes.
If `FALSE`, relations defined as unrelated (based on the relatedness thresholds) will not be saved.
If `NULL` (the default), the parameter will be set to `FALSE` for sample sizes &gt;= 1000, and to `TRUE` otherwise.</p></td>
    </tr>
    <tr>
      <th>path_failed_measurements</th>
      <td><p>Filepath (.rds extension) for saving a matrix that indicates for each measurement whether it failed
based on the detP and beadNr cutoffs. Defaults to `NULL`, in which case this matrix is not saved.
This option is useful for the <code><a href='mask_measurements.html'>mask_measurements</a></code> function,
which sets failed measurements in a beta-matrix to missing.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>Number of cores to use, defaults to 1.
The `BPPARAM` argument controls parallelization (see: <a href='https://bioconductor.org/packages/devel/bioc/vignettes/BiocParallel/inst/doc/Introduction_To_BiocParallel.pdf'>BiocParallel</a>).</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>Relevant if the `cores` argument &gt; 1.
Specifies the parallel backend to use
(see <a href='https://bioconductor.org/packages/devel/bioc/vignettes/BiocParallel/inst/doc/Introduction_To_BiocParallel.pdf'>BiocParallel</a>),
defaults to <code><a href='https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html'>MulticoreParam</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Several arguments can be specified,
only the `samplesheet` parameter is mandatory.
Note that the samplesheet must contain the following columns: Sample_Name','Sex' (M/F) and 'Basename'.
The 'Basename' column should contain paths to the IDAT files.
Optionally, 'Age', 'Smoking', 'BMI' and 'Alcohol' columns can be included.
If these are provided provided, these can be compared to DNA-methylation-based predictors (PMSs)
of the respective variables in the <code><a href='QCapp.html'>QCapp</a></code>.
All other included columns are kept and can be used within the <code><a href='QCapp.html'>QCapp</a></code> to color the variables by those columns</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Load example data from minfiData package (6 samples)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>minfiData</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://dplyr.tidyverse.org'>dplyr</a></span><span class='op'>)</span>
<span class='va'>baseDir</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, package<span class='op'>=</span><span class='st'>"minfiData"</span><span class='op'>)</span>
<span class='va'>samplesheet</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/minfi/man/read.metharray.sheet.html'>read.metharray.sheet</a></span><span class='op'>(</span><span class='va'>baseDir</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>[read.metharray.sheet] Found the following CSV files:</span></div><div class='output co'>#&gt; [1] "/Library/Frameworks/R.framework/Versions/3.6/Resources/library/minfiData/extdata/SampleSheet.csv"</div><div class='input'><span class='co'># Make sure that the samplesheet contains a 'Sample_Name'</span>
<span class='co'># column with the sample IDs and a 'Sex' column!</span>
<span class='va'>samplesheet</span> <span class='op'>&lt;-</span> <span class='va'>samplesheet</span> <span class='op'>%&gt;%</span>
                   <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>Sample_Name <span class='op'>=</span>  <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='va'>samplesheet</span><span class='op'>$</span><span class='va'>Slide</span>,
                                 <span class='va'>samplesheet</span><span class='op'>$</span><span class='va'>Array</span>, sep <span class='op'>=</span> <span class='st'>"_"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
                   <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/rename.html'>rename</a></span><span class='op'>(</span>Sex <span class='op'>=</span> <span class='va'>sex</span><span class='op'>)</span>

<span class='co'>## Get QC metrics</span>
<span class='va'>metrics</span> <span class='op'>&lt;-</span> <span class='fu'>get_qcmetrics</span><span class='op'>(</span><span class='va'>samplesheet</span>,
  name <span class='op'>=</span> <span class='st'>"example minfiData"</span>,
  calculate_PCs <span class='op'>=</span> <span class='cn'>TRUE</span>,
  n_probes_PCA <span class='op'>=</span> <span class='fl'>20000</span>,
  heatmap_variables <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Array"</span>, <span class='st'>"Slide"</span><span class='op'>)</span>,
  heatmap_variables_fct <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Array"</span>, <span class='st'>"Slide"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Creating a QCmetrics object</span>
#&gt; <span class='message'>Analysis started at: 2021-09-20 09:30:50</span></div><div class='output co'>#&gt; <span class='message'>Array: 450k</span>
#&gt; <span class='message'>-----------------------------</span></div><div class='output co'>#&gt; <span class='message'>Extracting QCmetrics in chunks of 40 samples..</span></div><div class='output co'>#&gt; <span class='message'>Using 144/144 probes for alcohol prediction</span></div><div class='output co'>#&gt; <span class='message'>Using 187/187 probes for smoking prediction</span></div><div class='output co'>#&gt; <span class='message'>Using 435/435 probes for BMI prediction</span></div><div class='output co'>#&gt; <span class='message'>Done!</span>
#&gt; <span class='message'>Merging the chunks..</span></div><div class='output co'>#&gt; <span class='message'>Performing IBS..</span></div><div class='output co'>#&gt; <span class='message'>10 out of 65 probes could be accurately called</span></div><div class='output co'>#&gt; <span class='message'>Calculating relatedness using the DNAm-inferred SNPs</span></div><div class='output co'>#&gt; <span class='message'>Performing array-wide PCA..</span></div><div class='output co'>#&gt; <span class='message'>Performing control probe PCA..</span></div><div class='output co'>#&gt; <span class='message'>Performing association tests between array-wide PCs and technical/biological variables..</span></div><div class='output co'>#&gt; <span class='message'>The following variables have zero variance, and will therefore be excluded from the association tests:</span></div><div class='output co'>#&gt; <span class='message'>	CD8T</span></div><div class='output co'>#&gt; <span class='message'>	Neutro</span></div><div class='output co'>#&gt; <span class='message'>Performing association tests between control probe PCs and technical/biological variables..</span></div><div class='output co'>#&gt; <span class='message'>The following variables have zero variance, and will therefore be excluded from the association tests:</span></div><div class='output co'>#&gt; <span class='message'>	CD8T</span></div><div class='output co'>#&gt; <span class='message'>	Neutro</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Merging the metrics..</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Number of outliers using default thresholds:</span></div><div class='output co'>#&gt; <span class='message'>MU: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>RG_ratio: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>GR_ratio: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>OP: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>HC: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>bscon: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>detectionP: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span></div><div class='output co'>#&gt; <span class='message'>beadNr: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>sex: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>Total: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span></div><div class='output co'>#&gt; <span class='message'>Done!</span></div><div class='output co'>#&gt; <span class='message'>Analysis finished at: 2021-09-20 09:31:20</span>
#&gt; <span class='message'></span></div><div class='input'>
<span class='co'># Add some custom thresholds:</span>
<span class='co'># Note that you can specify only the thresholds that you want to change,</span>
<span class='co'># for the others the default values will be used</span>
 <span class='va'>thresholds</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>MU <span class='op'>=</span> <span class='fl'>2000</span>, OP <span class='op'>=</span> <span class='fl'>13</span>, RG_ratio <span class='op'>=</span> <span class='fl'>0.6</span><span class='op'>)</span>
 <span class='va'>metrics</span> <span class='op'>&lt;-</span> <span class='fu'>get_qcmetrics</span><span class='op'>(</span><span class='va'>samplesheet</span>,
  name <span class='op'>=</span> <span class='st'>"example minfiData"</span>,
  calculate_PCs <span class='op'>=</span> <span class='cn'>TRUE</span>,
  n_probes_PCA <span class='op'>=</span> <span class='fl'>20000</span>,
  sample_qc_thresholds <span class='op'>=</span> <span class='va'>thresholds</span>,
  heatmap_variables <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Array"</span>, <span class='st'>"Slide"</span><span class='op'>)</span>,
  heatmap_variables_fct <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Array"</span>, <span class='st'>"Slide"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Creating a QCmetrics object</span>
#&gt; <span class='message'>Analysis started at: 2021-09-20 09:31:20</span></div><div class='output co'>#&gt; <span class='message'>Array: 450k</span>
#&gt; <span class='message'>-----------------------------</span></div><div class='output co'>#&gt; <span class='message'>Not all sample QC thresholds provided, defaults will be used for the following thresholds:</span></div><div class='output co'>#&gt; <span class='message'>GR_ratio</span></div><div class='output co'>#&gt; <span class='message'>HC</span></div><div class='output co'>#&gt; <span class='message'>bscon</span></div><div class='output co'>#&gt; <span class='message'>detP</span></div><div class='output co'>#&gt; <span class='message'>beadNr</span></div><div class='output co'>#&gt; <span class='message'>IBS_mean</span></div><div class='output co'>#&gt; <span class='message'>IBS_var</span></div><div class='output co'>#&gt; <span class='message'>XY_diff</span></div><div class='output co'>#&gt; <span class='message'>custom_outliers</span></div><div class='output co'>#&gt; <span class='message'>Extracting QCmetrics in chunks of 40 samples..</span></div><div class='output co'>#&gt; <span class='message'>Using 144/144 probes for alcohol prediction</span></div><div class='output co'>#&gt; <span class='message'>Using 187/187 probes for smoking prediction</span></div><div class='output co'>#&gt; <span class='message'>Using 435/435 probes for BMI prediction</span></div><div class='output co'>#&gt; <span class='message'>Done!</span>
#&gt; <span class='message'>Merging the chunks..</span></div><div class='output co'>#&gt; <span class='message'>Performing IBS..</span></div><div class='output co'>#&gt; <span class='message'>10 out of 65 probes could be accurately called</span></div><div class='output co'>#&gt; <span class='message'>Calculating relatedness using the DNAm-inferred SNPs</span></div><div class='output co'>#&gt; <span class='message'>Performing array-wide PCA..</span></div><div class='output co'>#&gt; <span class='message'>Performing control probe PCA..</span></div><div class='output co'>#&gt; <span class='message'>Performing association tests between array-wide PCs and technical/biological variables..</span></div><div class='output co'>#&gt; <span class='message'>The following variables have zero variance, and will therefore be excluded from the association tests:</span></div><div class='output co'>#&gt; <span class='message'>	CD8T</span></div><div class='output co'>#&gt; <span class='message'>	Neutro</span></div><div class='output co'>#&gt; <span class='message'>Performing association tests between control probe PCs and technical/biological variables..</span></div><div class='output co'>#&gt; <span class='message'>The following variables have zero variance, and will therefore be excluded from the association tests:</span></div><div class='output co'>#&gt; <span class='message'>	CD8T</span></div><div class='output co'>#&gt; <span class='message'>	Neutro</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Merging the metrics..</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Number of outliers using user-defined thresholds:</span></div><div class='output co'>#&gt; <span class='message'>MU: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>RG_ratio: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>GR_ratio: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>OP: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>HC: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>bscon: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>detectionP: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span></div><div class='output co'>#&gt; <span class='message'>beadNr: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>sex: 0 outliers (0%)</span></div><div class='output co'>#&gt; <span class='message'>Total: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span></div><div class='output co'>#&gt; <span class='message'>Done!</span></div><div class='output co'>#&gt; <span class='message'>Analysis finished at: 2021-09-20 09:31:50</span>
#&gt; <span class='message'></span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Paul Hop.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


