<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create a QC-metrics file — getQCmetrics • QCpG</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create a QC-metrics file — getQCmetrics"><meta property="og:description" content="Extract QC metrics and outlier information from IDAT files.
Returns summarized QC metrics, sample and probe outliers, relatedness info and
DNA-methylation-based predictors (PMSs) of various traits/exposures
in a QCmetrics object, which can be queried interactively using the QCapp method."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QCpG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/GettingStarted.html">Getting Started</a>
</li>
<li>
  <a href="../articles/UsingTheApp.html">Using the QCapp</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    In-depth articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GetExampleData.html">Get example data</a>
    </li>
    <li>
      <a href="../articles/AdvancedOptions.html">Advanced Options</a>
    </li>
    <li>
      <a href="../articles/EWASworkflow.html">EWAS workflow</a>
    </li>
    <li>
      <a href="../articles/AvailableMetrics.html">Available Metrics</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://twitter.com/pj_hop" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/pjhop/QCpG" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a QC-metrics file</h1>
    
    <div class="hidden name"><code>getQCmetrics.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract QC metrics and outlier information from IDAT files.
Returns summarized QC metrics, sample and probe outliers, relatedness info and
DNA-methylation-based predictors (PMSs) of various traits/exposures
in a <a href="QCmetrics-class.html">QCmetrics</a> object, which can be queried interactively using the <code><a href="QCapp.html">QCapp</a></code> method.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getQCmetrics</span><span class="op">(</span>
  <span class="va">samplesheet</span>,
  name <span class="op">=</span> <span class="cn">NULL</span>,
  experimenter <span class="op">=</span> <span class="cn">NULL</span>,
  chunk_size <span class="op">=</span> <span class="fl">40</span>,
  geno <span class="op">=</span> <span class="cn">NULL</span>,
  snplinker <span class="op">=</span> <span class="cn">NULL</span>,
  logfile <span class="op">=</span> <span class="cn">NULL</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  probe_qc_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"post_sampleqc"</span>, <span class="st">"pre_sampleqc"</span><span class="op">)</span>,
  heatmap_variables <span class="op">=</span> <span class="cn">NULL</span>,
  heatmap_variables_fct <span class="op">=</span> <span class="cn">NULL</span>,
  sample_qc_thresholds <span class="op">=</span> <span class="cn">NULL</span>,
  probe_qc_thresholds <span class="op">=</span> <span class="cn">NULL</span>,
  detp_threshold <span class="op">=</span> <span class="fl">1e-16</span>,
  beadnr_threshold <span class="op">=</span> <span class="fl">3</span>,
  relatedness_thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>IBS_mean <span class="op">=</span> <span class="fl">1.5</span>, IBS_var <span class="op">=</span> <span class="fl">0.35</span><span class="op">)</span>,
  custom_outliers <span class="op">=</span> <span class="cn">NULL</span>,
  epidish_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RPC"</span>, <span class="st">"CBS"</span>, <span class="st">"CP"</span><span class="op">)</span>,
  tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blood"</span>, <span class="st">"breast"</span>, <span class="st">"epithelium"</span><span class="op">)</span>,
  calculate_PCs <span class="op">=</span> <span class="cn">FALSE</span>,
  n_probes_PCA <span class="op">=</span> <span class="cn">NULL</span>,
  n_estimate_PCA <span class="op">=</span> <span class="va">chunk_size</span>,
  nrPCs <span class="op">=</span> <span class="fl">10</span>,
  save_all_relationships <span class="op">=</span> <span class="cn">NULL</span>,
  path_failed_measurements <span class="op">=</span> <span class="cn">NULL</span>,
  cores <span class="op">=</span> <span class="fl">1</span>,
  BPPARAM <span class="op">=</span> <span class="cn">NULL</span>,
  force <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>samplesheet</dt>
<dd><p>Samplesheet. Should contain the following columns: Sample_Name','Sex' (M/F) and 'Basename'.</p></dd>
<dt>name</dt>
<dd><p>Optionally, provide a name for the experiment (will be included in metadata).</p></dd>
<dt>experimenter</dt>
<dd><p>Optionally, name of the experiment (will be included in metadata).</p></dd>
<dt>chunk_size</dt>
<dd><p>Size of chunks, default = 40. Larger chunks will be a little bit faster
but will require more RAM.</p></dd>
<dt>geno</dt>
<dd><p>An optional matrix with genotypes (rows = snps, columns = samples)
to check for concordance between DNAm and genotype data. See the tutoriols for more information.</p></dd>
<dt>snplinker</dt>
<dd><p>Two-column data frame that links DNAm probes to genotype probes (first column = DNAm probes, second column = genotype IDs).
If not provided the 65 designated SNP-probes will be used.</p></dd>
<dt>logfile</dt>
<dd><p>Filepath to write logfile to, by default no logfile is written (`logfile = NULL`).</p></dd>
<dt>verbose</dt>
<dd><p>If `TRUE` (the default), information is printed in the console.</p></dd>
<dt>probe_qc_method</dt>
<dd><p>Should probe QC be performed after sample QC ('post_sampleqc') or prior to sample QC ('pre_sampleqc')?.
If 'post_sampleqc', sample outliers will be removed before performing site QC.</p></dd>
<dt>heatmap_variables</dt>
<dd><p>Character vector specifying additional variables to test for association with PCs (default = `NULL`).</p></dd>
<dt>heatmap_variables_fct</dt>
<dd><p>Relevant if the `heatmap_variables` argument is not `NULL`.
Which of the variables specified in `heatmap_variables` should be treated as factors?</p></dd>
<dt>sample_qc_thresholds</dt>
<dd><p>List of sample QC thresholds, if `NULL` (the default),
default thresholds will be used.
Note that custom thresholds can be specified for a subset of metrics, default
thresholds will then be used for metrics that are not specified.</p></dd>
<dt>probe_qc_thresholds</dt>
<dd><p>List of probe QC thresholds, if `NULL` (the default),
default thresholds will be used.</p></dd>
<dt>detp_threshold</dt>
<dd><p>detection P-value threshold (default = 1e-16).</p></dd>
<dt>beadnr_threshold</dt>
<dd><p>beadnr threshold (default = 3).</p></dd>
<dt>relatedness_thresholds</dt>
<dd><p>Initial relatedness thresholds, can be modified in the <code><a href="QCapp.html">QCapp</a></code>.</p></dd>
<dt>custom_outliers</dt>
<dd><p>Optional additional column name(s) in the samplesheet that
that indicate that samples should be treated as outliers.
The indicated columns should be of class `logical`, (`TRUE` = outlier).</p></dd>
<dt>epidish_method</dt>
<dd><p>Which method to use in the EpiDish <a href="https://rdrr.io/pkg/EpiDISH/man/epidish.html" class="external-link">epidish</a> cellcount prediction (RPC, CBS, or CP).
Default = 'RPC'.</p></dd>
<dt>tissue</dt>
<dd><p>Tissue-type for cellcount-estimation.
Currently 'blood', 'breast' and 'epithelium' are implemented, following the <a href="https://rdrr.io/pkg/EpiDISH/man/epidish.html" class="external-link">epidish</a> function.</p></dd>
<dt>calculate_PCs</dt>
<dd><p>Should PCA be performed? (TRUE/FALSE).
Default = `FALSE`. For large sample sizes RAM could become a bottleneck.
To limit RAM usage (besides setting this parameter to `FALSE` to skip PCA), the
`n_probes_PCA` argument can be used to limit PCA to a subset of most variable probes.</p></dd>
<dt>n_probes_PCA</dt>
<dd><p>Number of most variable probes to maintain for PCA,
defaults to 'NULL' (all probes are used).</p></dd>
<dt>n_estimate_PCA</dt>
<dd><p>Relevant if `n_probes_PCA` is not `NULL`.
This argument specifies the number of samples used to estimate the most variable probes (defaults to 'chunk_size').</p></dd>
<dt>nrPCs</dt>
<dd><p>nr of PCs to retain. Default = 10</p></dd>
<dt>save_all_relationships</dt>
<dd><p>Save all relatedness data (IBS)?
Relatedness output contains N^2 rows, and thus it will rapidly increase in size with large sample sizes.
If `FALSE`, relations defined as unrelated (based on the relatedness thresholds) will not be saved.
If `NULL` (the default), the parameter will be set to `FALSE` for sample sizes &gt;= 1000, and to `TRUE` otherwise.</p></dd>
<dt>path_failed_measurements</dt>
<dd><p>Filepath (.rds extension) for saving a matrix that indicates for each measurement whether it failed
based on the detP and beadNr cutoffs. Defaults to `NULL`, in which case this matrix is not saved.
This option is useful for the <code><a href="mask_measurements.html">mask_measurements</a></code> function,
which sets failed measurements in a beta-matrix to missing.</p></dd>
<dt>cores</dt>
<dd><p>Number of cores to use, defaults to 1.
The `BPPARAM` argument controls parallelization (see: <a href="https://bioconductor.org/packages/devel/bioc/vignettes/BiocParallel/inst/doc/Introduction_To_BiocParallel.pdf" class="external-link">BiocParallel</a>).</p></dd>
<dt>BPPARAM</dt>
<dd><p>Relevant if the `cores` argument &gt; 1.
Specifies the parallel backend to use
(see <a href="https://bioconductor.org/packages/devel/bioc/vignettes/BiocParallel/inst/doc/Introduction_To_BiocParallel.pdf" class="external-link">BiocParallel</a>),
defaults to <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></code></p></dd>
<dt>force</dt>
<dd><p>See the <a href="https://rdrr.io/pkg/minfi/man/minfi-package.html" class="external-link">minfi</a> documentation: Should reading different size IDAT files be forced?
Defaults to FALSE.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <a href="QCmetrics-class.html">QCmetrics</a> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Although many parameters can be tweaked,
only the `samplesheet` parameter is mandatory.
Note that the samplesheet must contain the following columns: Sample_Name','Sex' (M/F) and 'Basename'.
The 'Basename' column should contain paths to the IDAT files.
Optionally, 'Age', 'Smoking', 'BMI' and 'Alcohol' columns can be included.
If these are provided provided, these can be compared to DNA-methylation-based predictors (PMSs)
of the respective variables in the <code><a href="QCapp.html">QCapp</a></code>.
All other included columns are kept and can be used within the <code><a href="QCapp.html">QCapp</a></code> to color the variables by those columns</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Load example data from minfiData package (6 samples)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">minfiData</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: minfi</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: DelayedArray</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocParallel</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘DelayedArray’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     aperm, apply, rowsum</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biostrings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: XVector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biostrings’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     strsplit</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: bumphunter</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: foreach</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: iterators</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: locfit</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> locfit 1.5-9.4 	 2020-03-24</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IlluminaHumanMethylation450kmanifest</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IlluminaHumanMethylation450kanno.ilmn12.hg19</span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘dplyr’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:minfi’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     combine</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biostrings’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     collapse, intersect, setdiff, setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:XVector’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     slice</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     count</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     combine</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:GenomicRanges’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect, setdiff, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:GenomeInfoDb’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:IRanges’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     collapse, desc, intersect, setdiff, slice, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:S4Vectors’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     first, intersect, rename, setdiff, setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:BiocGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     combine, intersect, setdiff, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     filter, lag</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect, setdiff, setequal, union</span>
<span class="r-in"><span class="va">baseDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"minfiData"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">samplesheet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/minfi/man/read.metharray.sheet.html" class="external-link">read.metharray.sheet</a></span><span class="op">(</span><span class="va">baseDir</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [read.metharray.sheet] Found the following CSV files:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/hpc/hers_en/phop/software/R-3.6.3/libs/minfiData/extdata/SampleSheet.csv"</span>
<span class="r-in"><span class="co"># Make sure that the samplesheet contains a 'Sample_Name'</span></span>
<span class="r-in"><span class="co"># column with the sample IDs and a 'Sex' column!</span></span>
<span class="r-in"><span class="va">samplesheet</span> <span class="op">&lt;-</span> <span class="va">samplesheet</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">                   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Sample_Name <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Slide</span>,</span>
<span class="r-in">                                 <span class="va">samplesheet</span><span class="op">$</span><span class="va">Array</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span>
<span class="r-in">                   <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>Sex <span class="op">=</span> <span class="va">sex</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Get QC metrics</span></span>
<span class="r-in"><span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu">getQCmetrics</span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span class="r-in">  name <span class="op">=</span> <span class="st">"example minfiData"</span>,</span>
<span class="r-in">  calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">  n_probes_PCA <span class="op">=</span> <span class="fl">20000</span>,</span>
<span class="r-in">  heatmap_variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Array"</span>, <span class="st">"Slide"</span><span class="op">)</span>,</span>
<span class="r-in">  heatmap_variables_fct <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Array"</span>, <span class="st">"Slide"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating a QCmetrics object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysis started at: 2022-09-27 08:55:49</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Array: 450k</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -----------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting QCmetrics in chunks of 40 samples..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No methods found in package ‘RSQLite’ for request: ‘dbListFields’ when loading ‘lumi’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 144/144 probes for alcohol prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 187/187 probes for smoking prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 435/435 probes for BMI prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging the chunks..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing IBS..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 out of 65 probes could be accurately called</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating relatedness using the DNAm-inferred SNPs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing array-wide PCA..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing control probe PCA..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing association tests between array-wide PCs and technical/biological variables..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following variables have zero variance, and will therefore be excluded from the association tests:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	CD8T</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	Neutro</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing association tests between control probe PCs and technical/biological variables..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following variables have zero variance, and will therefore be excluded from the association tests:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	CD8T</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	Neutro</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging the metrics..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of outliers using default thresholds:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MU: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> RG_ratio: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> GR_ratio: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OP: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> HC: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bscon: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> detectionP: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> beadNr: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> sex: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysis finished at: 2022-09-27 08:56:49</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Add some custom thresholds:</span></span>
<span class="r-in"><span class="co"># Note that you can specify only the thresholds that you want to change,</span></span>
<span class="r-in"><span class="co"># for the others the default values will be used</span></span>
<span class="r-in"> <span class="va">thresholds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>MU <span class="op">=</span> <span class="fl">2000</span>, OP <span class="op">=</span> <span class="fl">13</span>, RG_ratio <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span>
<span class="r-in"> <span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu">getQCmetrics</span><span class="op">(</span><span class="va">samplesheet</span>,</span>
<span class="r-in">  name <span class="op">=</span> <span class="st">"example minfiData"</span>,</span>
<span class="r-in">  calculate_PCs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">  n_probes_PCA <span class="op">=</span> <span class="fl">20000</span>,</span>
<span class="r-in">  sample_qc_thresholds <span class="op">=</span> <span class="va">thresholds</span>,</span>
<span class="r-in">  heatmap_variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Array"</span>, <span class="st">"Slide"</span><span class="op">)</span>,</span>
<span class="r-in">  heatmap_variables_fct <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Array"</span>, <span class="st">"Slide"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating a QCmetrics object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysis started at: 2022-09-27 08:56:49</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Array: 450k</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -----------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Not all sample QC thresholds provided, defaults will be used for the following thresholds:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> GR_ratio</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> HC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bscon</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> detP</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> beadNr</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> IBS_mean</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> IBS_var</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> XY_diff</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> custom_outliers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting QCmetrics in chunks of 40 samples..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 144/144 probes for alcohol prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 187/187 probes for smoking prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 435/435 probes for BMI prediction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging the chunks..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing IBS..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10 out of 65 probes could be accurately called</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating relatedness using the DNAm-inferred SNPs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing array-wide PCA..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing control probe PCA..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing association tests between array-wide PCs and technical/biological variables..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following variables have zero variance, and will therefore be excluded from the association tests:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	CD8T</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	Neutro</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing association tests between control probe PCs and technical/biological variables..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following variables have zero variance, and will therefore be excluded from the association tests:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	CD8T</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	Neutro</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging the metrics..</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of outliers using user-defined thresholds:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MU: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> RG_ratio: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> GR_ratio: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OP: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> HC: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> bscon: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> detectionP: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> beadNr: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> sex: 0 outliers (0%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total: 2 outliers (33.33%) [WARNING] High percentage of outliers!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Analysis finished at: 2022-09-27 08:57:27</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Paul Hop.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

